csharp/test/Ice/metrics/AllTests.cs

```diff
@@ -1109,7 +1109,7 @@ public static IMetricsPrx Run(TestHelper helper, CommunicatorObserver obsv, bool
             CheckFailure(clientMetrics, "Invocation", im1.Id, "System.Exception", 2, output);
 
             im1 = (InvocationMetrics)map["opWithRequestFailedException"];
-            TestHelper.Assert(im1.Current <= 1 && im1.Total == 2 && im1.Retry == 0);
+            TestHelper.Assert(im1.Current <= 1 && im1.Total == 2 && im1.Retry == 2);
```

Comment:
there are two calls each has a retry, the retry fails with NoEndpointException. As ONE here is thrown from user code we can use NoRetry policy and get 0 retries, but currently is thrown with OtherReplica.

Reply:
For example, here we don't know if there are more endpoints until we make the new attempt, so if you get OtherReplica you will attempt and it can fail with NoEndpointException if all known replicas have been already tried

Reply:
So, the client gets a remote exception RetryPolicy=OtherReplica (it's ONE but the client doesn't unmarshal it) for the first attempt (retry #0). The proxy is a single-proxy endpoint. I don't see where the next 2 attempts come from.

Also, naturally, if my first attempt throws SomeException, and the retry fails with NoEndpointException (because there is no other endpoint to try), I want to pass SomeException to the application, and not NoEndpointException.

Reply:
Here the test calls twice opWithRequestFailedException so Total == 2 , and one retry per invocation so Retry = 2. The metrics just show the accumulated result for the operation.

When we get OtherReplica we don't know if there are more endpoints to try, we just add the connector to the excluded list so that is no longer used for the current retry sequence and try again. The next attempt call GetConnectionAsync that throws NoEndpointException, and NoEndpointException handler breaks the retry loop that will return the last failure ONE in this case.

Reply:
Ok, I see now, thanks!